### PR TITLE
Support c++17 & Slight Improvement

### PR DESCRIPTION
**Pull request by dongho-Han**
This is a PR for supporting c++17.
~~Additional solved issues:~~
  - ~~Regarding issue #268: change the test model~~
  - ~~Fix issue #312~~ 
  - ~~Deprecate `resource_monitor_test`~~
  
  Additional Improvements
- `.bazelrc`: use `common` to use cache betw `build` and `test`
- In "cpu_test" & "resource_monitor_test": check the device's status using `is_rooted`

Tested with these environments:
  - Android S22: Unrooted
  - Google Pixel 4: Rooted
  - Linux(Server, Local)
  - Windows(thanks to @dostos)

### CODE REVIEW

File: band/model_analyzer.cc [hunk 1]
```diff
@@ -351,12 +351,11 @@ absl::Status ModelAnalyzer::GetUnitSubgraphs(
           continue;
         }
 
-        WorkerId tmp_worker_id = static_cast<int>(device_flag);
-        if (unsupported_ops.find(tmp_worker_id) == unsupported_ops.end() ||
-            unsupported_ops.at(tmp_worker_id).find(op_index) ==
-                unsupported_ops.at(tmp_worker_id).end()) {
-          if (op_sets_to_ignore[tmp_worker_id].find(op_index) ==
-              op_sets_to_ignore[tmp_worker_id].end()) {
+        if (unsupported_ops.find(worker_id) == unsupported_ops.end() ||
```
Comment (dostos) [inline band/model_analyzer.cc:354]: Does this fix is relevant to the C++17 support? Please use separate PR for independent fixes next time :)

File: band/test/BUILD [hunk 2]
```diff
@@ -159,16 +159,17 @@ band_cc_android_test(
...
-    name = "resource_monitor_test",
-    size = "small",
-    srcs = ["resource_monitor_test.cc"],
-    data = [],
-    deps = [
-        "//band:resource_monitor",
-        "@com_google_googletest//:gtest",
-    ],
-)
+# # Deprecated
```
Comment (dostos) [inline band/test/BUILD:162]: Now we could uncomment this?

File: band/test/cpu_test.cc [hunk 3]
```diff
@@ -28,20 +29,25 @@ namespace test {
...
-    if (!set_status.ok()) {
-      EXPECT_EQ(target_set.NumEnabled(), 0);
-    } else {
-      sleep(1);
-      CpuSet current_set;
-      // should always success
-      EXPECT_EQ(GetCPUThreadAffinity(current_set), absl::OkStatus());
-      EXPECT_EQ(target_set, current_set);
+  if (!device::IsRooted()) {
+      std::cout << "As this device is unrooted, Cpu Affinity can't be set.\n" << std::endl;
```
Comment (dostos) [inline band/test/cpu_test.cc:33]:
```suggestion
      GTEST_SKIP() << "As this device is unrooted, Cpu Affinity can't be set.\n" << std::endl;
```
Better solution : https://google.github.io/googletest/advanced.html#skipping-test-execution.

Comment (dostos) [reply band/test/cpu_test.cc:33]: JFYI, it's not good to print something during the test.

File: band/test/resource_monitor_test.cc [hunk 4]
```diff
@@ -143,73 +144,78 @@ TEST(ResourceMonitorTest, GetCpuFreqTest) {
...
-    EXPECT_EQ(
-        monitor.AddCpuFreqResource(flag, CpuFreqFlag::DOWN_TRANSITION_LATENCY),
-        absl::OkStatus());
-    // This is optional. E.g., Pixel 4 doesn't have this.
-    auto status =
-        monitor.AddCpuFreqResource(flag, CpuFreqFlag::TRANSITION_COUNT);
-    std::cout << "AddCpuFreqResource " << ToString(flag)
-              << " TRANSITION_COUNT: " << status << std::endl;
+  if (!device::IsRooted()) {
+      std::cout << "As this device is unrooted, CpuFreqFlag can't be adapted.\n" << std::endl;
```
Comment (dostos) [inline band/test/resource_monitor_test.cc:148]: Same as above

Comment (dostos) [reply band/test/resource_monitor_test.cc:148]: \n << std::endl; will result two newlines

File: script/utils/util.py [hunk 5]
```diff
@@ -109,14 +109,14 @@ def get_dst_path(base_dir, target_platform, debug):
 
 
 def push_to_android(src, dst):
-    run_cmd(f'adb -d push {src} {ANDROID_BASE}{dst}')
+    run_cmd(f'adb push {src} {ANDROID_BASE}{dst}')
```
Comment (dostos) [inline script/utils/util.py:112]: Why should we delete these flags? "-d" is necessary to specify the physical device if adb detects it along with android simulators. 

It also seems irrelevant to the PR.

File: script/run_test.py [hunk 6]
```diff
@@ -37,6 +37,7 @@ def get_target(debug, postfix=""):
         clean_bazel(args.docker)
     
     if args.android:
```
Comment (dostos) [inline script/run_test.py:39]: Why should we wipe the base dir?

File: script/run_test.py [hunk 7]
```diff
@@ -69,7 +70,7 @@ def get_target(debug, postfix=""):
 
         print("Clean up test directory from device")
         run_cmd(
-            f'adb -d shell rm -r /data/local/tmp/{temp_dir_name}')
+            f'adb shell rm -r /data/local/tmp/{temp_dir_name}')
```
Comment (dostos) [inline script/run_test.py:73]: same as below

**Review by dostos — APPROVED** [review]
Thank you for your tremendous efforts 👍🏼 
Looks good to me except for a few minor comments.

File: .bazelrc [hunk 8]
```diff
@@ -2,19 +2,34 @@
...
+build:windows --cxxopt=/std:c++17
+build:windows --host_cxxopt=/std:c++17
+test:linux --cxxopt=-std=c++17
+test:linux --host_cxxopt=-std=c++17
+test:android --cxxopt=-std=c++17
+test:android --host_cxxopt=-std=c++17
+test:windows --cxxopt=/std:c++17
+test:windows --host_cxxopt=/std:c++17
+
+# Deprecated
```
[the remark this answers is not in the document]
Comment (dongho-Han) [reply .bazelrc:19]: I'll remove all the deprecated

File: band/test/BUILD (hunk 2 above)
Comment (dongho-Han) [reply band/test/BUILD:162]: I think deleting `resource_monitor_test` test is not relevant to C++17 PR. As you mentioned, how about making a seperate PR for this and issue #312 problem as solved in `band/model_analyzer.cc`?

File: band/test/cpu_test.cc (hunk 3 above)
Comment (dongho-Han) [reply band/test/cpu_test.cc:33]: I'll update as not printing something.

File: script/run_test.py (hunk 6 above)
Comment (dongho-Han) [reply script/run_test.py:39]: I updated that line because when we run test many times, build is failed because the built file already exists.
To solve this issue, I think several ways are possible:
- remove the build file all the times
- change this function usage with exist_ok=True: https://github.com/mrsnu/band/blob/e265b398aecda93294d37ebbae3d06f4ffa98794/script/utils/util.py#L66

However, with second way, as the `shutil.copytree` doesn't support overwrite, I think using `disutils.dir_util.copy_tree` is needed for multiple build test.
How's your opinion about that?

Comment (dongho-Han) [reply script/run_test.py:39]: FYI: I left the logs 
```
Traceback (most recent call last):
  File "script/run_test.py", line 55, in <module>
    copy(f'bazel-bin/band/test', get_target(args.debug))
  File "/workspaces/band/script/utils/util.py", line 66, in copy
    shutil.copytree(src, dst, ignore_dangling_symlinks=True, symlinks=True)
  File "/usr/lib/python3.8/shutil.py", line 557, in copytree
    return _copytree(entries=entries, src=src, dst=dst, symlinks=symlinks,
  File "/usr/lib/python3.8/shutil.py", line 458, in _copytree
    os.makedirs(dst, exist_ok=dirs_exist_ok)
  File "/usr/lib/python3.8/os.py", line 223, in makedirs
    mkdir(name, mode)
FileExistsError: [Errno 17] File exists: 'test_bin/armv8-a/release/test'
```

File: script/utils/util.py (hunk 5 above)
Comment (dongho-Han) [reply script/utils/util.py:112]: I deleted the `-d` option because it **doesn't support wireless connection** with adb. It is for the usb connection.
Do you think using `-d` option is more necessary?

File: band/test/BUILD (hunk 2 above)
Comment (dostos) [reply band/test/BUILD:162]: Sounds good!

File: script/utils/util.py (hunk 5 above)
Comment (dostos) [reply script/utils/util.py:112]: I see, maybe we should add additional option dedicated for any flags at the end of the ArgParse.

Comment (dongho-Han) [reply script/utils/util.py:112]: I'll separately handle this issue in another PR.

File: script/run_test.py (hunk 6 above)
Comment (dongho-Han) [reply script/run_test.py]: I'll separately handle this issue in another PR.

File: band/engine.cc
[the remark this answers is not in the document]
Comment (dongho-Han) [reply band/engine.cc:1]: Updated.

File: .bazelrc
[the remark this answers is not in the document]
Comment (dongho-Han) [reply .bazelrc:1]: @widiba03304 I reflected all the comments you suggested!
Could you take a look?

[the remark this answers is not in the document]
Comment (dongho-Han) [reply .bazelrc:1]: @widiba03304 
Could you check the latest commits?
I reflected all the suggestions!

[the remark this answers is not in the document]
Comment (widiba03304) [reply .bazelrc:1]: It looks all good to me!

**Comment by dongho-Han** [comment]
As I reflected all the comments, I will merge.
Thanks all of us!
